### PR TITLE
New time logic and a basic clock (+solved the bug of dark overlay during night)

### PR DESCRIPTION
**Pull request by Manunz15**
I refactored the game time logic by creating a new class called **GameTime** in _game_time.py_ to track time flow, replacing the previous system where **Sky** managed this. Additionally, I implemented a new logic that allows **Sky** to set its color based on the current game time. This means that after sleeping there is no need to reset the sky color, simply change the hour. I also developed a basic clock (**Clock** class in _overlay.py_) with two modes: analog and digital. These are simple designs, but I believe they can  serve as a good starting point for future developments.

Screenshot 2024-07-05 224745

Screenshot 2024-07-05 224819

I fixed a minor bug where the overlay became dark during the night just like the rest of the game.

Screenshot 2024-07-05 224925

Screenshot 2024-07-05 224852

Edit:

Finally I created a more flexible sky's color logic. Now it's based on a dictionary where you put {'hour': (color tuple)}, so we can put more colors during the day. These are screenshot where I put white at 12:00, red at 20:00, blue at 24:00 and light cyan at 6:00.

Screenshot 2024-07-07 002601
Screenshot 2024-07-07 002626
Screenshot 2024-07-07 002636
Screenshot 2024-07-07 002658

### CODE REVIEW

File: src/overlay.py
Comment (fortwoone) [inline src/overlay.py:120]: All this could be replaced by a single font.render and blit call. You might want to use f-strings instead when rendering the text

File: src/sky.py
Comment (fortwoone) [inline src/sky.py:36]: `list(self.color.keys())` is redundant. `list(self.colors)` will achieve the same effect. Also, you might want to use an equality comparison instead of using indexes here.

File: src/sky.py
Comment (fortwoone) [inline src/sky.py:47]: No need to regenerate both of these lists twice, nothing prevents you from storing them in a variable. In fact, that might be better, because of algorithmic complexity

File: src/sky.py
Comment (fortwoone) [inline src/sky.py:51]: Somehow, this feels like it can be simplified

**Review by fortwoone — CHANGES_REQUESTED** [review]
Overall, it's a good idea. But right now, the code structure is a bit messy and file/class names are confusing, so this will need to be reorganised before it's ready for merging. Also, some operations can be simplified, for the sake of optimisation (no need to iterate over the same list thrice, when you can just store the result somewhere).

File: src/game_time.py
Comment (Manunz15) [reply src/game_time.py:1]: You are right, I probably copied it for some reasons but forgot to delete it

File: src/overlay.py
Comment (Manunz15) [reply src/overlay.py:36]: I can't use enums, this PR was made before they were added, I will change this after merging

Comment (fortwoone) [reply src/overlay.py:36]: You need to fetch and pull from the main branch, otherwise the merge conflicts will get very high

File: src/overlay.py
Comment (Manunz15) [reply src/overlay.py:120]: I know I can use only one f-string, but I want to fix the position of ':', this way I obtained a way nicer clock were the text doesn't move based on the length of the text (digits have different pixels-width)

File: src/sky.py
Comment (Manunz15) [reply src/sky.py:51]: This is the simpler solution I was able to find

Comment (fortwoone) [reply src/sky.py:53]: `end_hour += 24 * (end_hour > start_hour)`

File: src/sky.py
Comment (Manunz15) [reply src/sky.py]: Solved

File: src/sky.py
Comment (Manunz15) [reply src/sky.py]: Solved

File: src/sky.py
Comment (fortwoone) [inline src/sky.py]: `self.colors_rgb` is already a list, you defined it so in the initialiser, so why convert it to a list again?

File: src/sky.py
Comment (fortwoone) [inline src/sky.py]: Slight oversight here: this will cause an error, because you forgot to put the parentheses when calling `self.colors.keys`. So right now you're actually calling the initialiser of the list class on _the bound method object representing `self.colors.keys`_, not the result of calling `self.colors.keys()`

File: src/sky.py
Comment (fortwoone) [inline src/sky.py]: Slight oopsie here: you can't subscript an int. You might have intended to do this instead:

```py
start_hour = int(self.colors_hours[color_index])
end_hour = int(self.colors_hours[color_index + 1])
```

**Review by fortwoone — CHANGES_REQUESTED** [review]
Found a few errors, but these are mostly due to slight lack of attention, and otherwise it seems fine

File: src/overlay.py
Comment (fortwoone) [reply src/overlay.py:120]: This could be fixed by editing the font itself, but I'm not sure we have the legal rights to do so

File: src/sky.py
Comment (Manunz15) [reply src/sky.py]: I think you reviewed outdated commits

Comment (Manunz15) [reply src/sky.py]: I already solved these problems in the 'Bug fix' commit

Comment (fortwoone) [reply src/sky.py]: I just saw you pushed a commit yes

File: src/sky.py
Comment (fortwoone) [reply src/sky.py:53]: This could be how you simplify that line (bools are derived from ints so this works)

**Review by fortwoone — COMMENTED** [review]
Just have one last little nitpick

File: src/sky.py
Comment (Manunz15) [reply src/sky.py]: Really good solution, I implemented it

File: main.py
Comment (fortwoone) [inline main.py:13]: This is not necessary anymore

**Review by fortwoone — COMMENTED** [review]
One little nitpick left

File: main.py
Comment (Manunz15) [reply main.py]: Yeah, I just forgot to delete it. Done

**Review by fortwoone — APPROVED** [review]

**Review by fortwoone — CHANGES_REQUESTED** [review]
I have to remove my approval now, because the codebase has been restructured and heavily changed in the past few days. As a result, a rebase onto main is required before I can give my approval again

**Pull request closed without merging**
